warehouse/forklift/legacy.py

```diff
@@ -571,6 +571,22 @@ def _is_valid_dist_file(filename, filetype):
     # allow it.
     return True
 
+def _is_duplicate_file(db_session, filename, hashes):
+    """ Check to see if file already exists, and if it's content matches
+    """
+
+    file_ = (
+        db_session.query(File)
+                  .filter(File.filename == filename)
+                  .first()
+    )
+
+    if file_:
+        return (file_.sha256_digest == hashes["sha256"] and
+                file_.md5_digest == hashes["md5"] and
+                file_.blake2_256_digest == hashes["blake2_256"])
+
+    return False
```

Comment:
Thinking this through, I wonder if this function shouldn't have 3 distinct return files, ``True``, ``False``, and ``None``. The return values would then mean:

* ``True``: This file is a duplicate and all further processing should halt.
* ``False``: This file exists, but it is not a duplicate.
* ``None``: This file does not exist.

Then further down we can do:

```python
is_duplicate =  _is_duplicate_file(request.db, filename, file_hashes):
if is_duplicate:
    raise _exc_with_message(HTTPBadRequest, "File already exists.")
elif is_duplicate is not None:
    return Response()
```

This means we get to skip a whole bunch of processing and uploading files to S3 etc for something that we know is a duplicate and we want to no-op anyways. It also prevents anything weird from happening like people lying to us about the contents of the file in order to change the data in the DB.

What do you think?

Reply:
I really like the idea of short-circuiting to prevent wasted calls/resources/etc. I'll write up a unit test for and see how it feels.

Reply:
I wrote up some unit tests for this function and modified it to support the above use case. Also, following the logic you suggested, this will now bail early if the file exists but isn't exactly the same. I'm not sure if we want to return a different response in that scenario or not.

Thoughts?

Reply:
You mean it will bail early if the file exists and *is* exactly the same, right?

Reply:
Right, so as it is written now the code will bail with an exception if a file exists and is exactly the same. 

This latest commit adds what you proposed above, which will bail early _without an error_ if the file exists, but not the same content. I could have totally misinterpreted your intent with that, so feel free to straighten me out.

Reply:
Oh right I see. Basically the goal is:

* If the file isn't the same file as we previously had, then bail early with an error. Since we already have a file we don't want to continue, but since this file doesn't match what was already uploaded we want to treat this as an error code.
* If the file is the same file as we previously had, just bail early with a success response. Since we've already had this file uploaded and it matches what was already uploaded we can just pretend we did the rest of the processing and it was successful.
* If we don't already have a file by this name, continue processing.

I might have switched around my conditional when I wrote the example code, but it sounds like we have the two first cases switched, so maybe we need to add a ``not`` to the return line of the function.

Reply:
Ah right, I'm following now. Making a sandwich then I'll fix this up.

Reply:
Feel like I have it covered. Take a look when you can.